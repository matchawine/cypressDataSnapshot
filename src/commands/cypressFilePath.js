//Use relative path because absolute has webpack problems
//Cannot use Cypress.spec when launching serveral tests, cf. https://github.com/cypress-io/cypress/issues/3090#issuecomment-889470707
const testFileRegExp = /\/(cypress\/.*)\/([^/]+)$/

const getTestFolderAndFileName = absoluteFile => {
  const [, testFolder, testFileName] = absoluteFile.match(testFileRegExp)
  return { testFolder, testFileName }
}

export const getTestPath = absoluteFile => {
  const { testFolder, testFileName } = getTestFolderAndFileName(absoluteFile)
  return `${testFolder}/${testFileName}`
}

export const getSnapshotFilePath = absoluteFile => {
  const { testFolder, testFileName } = getTestFolderAndFileName(absoluteFile)
  return `${testFolder}/__snapshots__/${testFileName}.snap`
}

export const getDefaultSnapshotName = Cypress =>
  Cypress.currentTest.titlePath.join(" ")
