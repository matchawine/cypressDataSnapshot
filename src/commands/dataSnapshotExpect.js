/**
 * Serializable Jest expect functions. Use it in dataSnapshot property matchers instead of expect.any(), etc...
 * Ex. instead of `expect.any(String)`, use  `dataSnapshotExpect("any", String)`
 *
 * @param func function path/name called from Jest expect. Ex. "any", "not.stringMatching"
 * @param args args to give the Jest function. Ex. `String`
 * @returns {{__isJestExpect: boolean, __args: *[], __function}}
 */
export const dataSnapshotExpect = (func, ...args) => ({
  __isJestExpect: true,
  __function: func,
  __args: args,
})
