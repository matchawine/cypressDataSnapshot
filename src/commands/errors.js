export const getUdateText = ({ testPath }) =>
  `To update all this spec snapshot, run:\n\`$ cypress run --env SNAPSHOT_UPDATE=all --spec ${testPath}\``

export const getFrontEndErrorText = ({
  snapshotName,
  testResult: { actual, expected },
  testPath,
}) =>
  `Snapshot name: \`${snapshotName}\`\n\nExpected: ${expected}\n\nActual: ${actual}\n\n${getUdateText(
    { testPath },
  )}`

export const logMessageInServer = ({
  testResult: { pass, message },
  testPath,
}) => {
  if (!pass) {
    console.warn(message())
    console.log(getUdateText({ testPath }))
  }
}
