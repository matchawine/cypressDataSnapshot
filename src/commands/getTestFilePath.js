const getAbsoluteFileFromCypressEvent = () =>
  cy.task("getTestFilePath").then(absoluteFileFromCypressEvent => {
    expect(absoluteFileFromCypressEvent).to.not.equal(undefined)
    return cy.wrap(absoluteFileFromCypressEvent)
  })

// Cannot use Cypress.spec when launching serveral tests, cf. https://github.com/cypress-io/cypress/issues/3090#issuecomment-889470707
// this.test.invocationDetails?.absoluteFile is flaky when runninbg `$ cypress run`. If it does not exist, we use an alternate method (that doesn't work in `$ cypress open`)
Cypress.Commands.add("getTestFilePath", function () {
  const absoluteFileFromInvocationDetails =
    this.test.invocationDetails?.absoluteFile
  return absoluteFileFromInvocationDetails
    ? cy.wrap(absoluteFileFromInvocationDetails)
    : getAbsoluteFileFromCypressEvent()
})
