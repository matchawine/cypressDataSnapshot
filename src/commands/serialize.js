import { serialize as originalSerialize } from "serialize-anything"

const nativeObjects = [
  "Date",
  "Number",
  "String",
  "Boolean",
  "Array",
  "RegExp",
  "Object",
]

/**
 * Serialize-javascript uses toString() and so sends back native constructors like `function String() { [native code]}`
 *
 * @param serializedValue
 * @returns {string}
 */
const cleanNativeObjectsInSerializedString = serializedValue =>
  nativeObjects.reduce(
    (serializedValue, nativeObject) =>
      serializedValue.replaceAll(
        `function ${nativeObject}() { [native code] }`,
        nativeObject,
      ),
    serializedValue,
  )

export const serialize = value => {
  const serializedValue = originalSerialize(value)
  return cleanNativeObjectsInSerializedString(serializedValue)
}
