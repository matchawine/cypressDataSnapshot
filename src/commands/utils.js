//Use relative path because absolute has webpack problems
//Cannot use Cypress.spec when launching serveral tests, cf. https://github.com/cypress-io/cypress/issues/3090#issuecomment-889470707
const testFileRegExp = /\/(cypress\/.*)\/([^/]+)$/

export const getSnapshotFilePath = absoluteFile => {
  const [, testFolder, testFileName] = absoluteFile.match(testFileRegExp)
  return `${testFolder}/__snapshots__/${testFileName}.snap`
}

export const getDefaultSnapshotName = Cypress =>
  Cypress.currentTest.titlePath.join(" ")

export const getErrorText = (testTitle, { actual, expected }) =>
  `Snapshot name: \`${testTitle}\`\n\nExpected: ${expected}\n\nActual: ${actual}`
