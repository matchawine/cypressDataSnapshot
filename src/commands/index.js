import {
  getDefaultSnapshotName,
  getErrorText,
  getSnapshotFilePath,
} from "./utils"
import { serialize } from "./serialize"

const getToMatchSnapshotCommand = ({ updateSnapshot }) =>
  function (actual, propertyMatchers, hint) {
    const snapshotFilePath = getSnapshotFilePath(
      this.test.invocationDetails.absoluteFile,
    )
    const snapshotName = getDefaultSnapshotName(Cypress)
    const serializedActual = serialize(actual)
    const serializedPropertyMatchers = serialize(propertyMatchers)

    cy.task("toMatchSnapshot", {
      serializedActual,
      snapshotFilePath,
      snapshotName,
      hint,
      serializedPropertyMatchers,
      updateSnapshot,
    }).then(testResult => {
      if (!testResult.pass)
        throw new Error(getErrorText(snapshotName, testResult))

      if (updateSnapshot)
        throw new Error(
          `Snapshot \`${snapshotName}\` updated!\n\nNow replace back the \`.updateSnapshot\` command with \`.toMatchSnapshot\``,
        )
    })
  }

Cypress.Commands.add(
  "toMatchSnapshot",
  getToMatchSnapshotCommand({ updateSnapshot: false }),
)

Cypress.Commands.add(
  "updateSnapshot",
  getToMatchSnapshotCommand({ updateSnapshot: true }),
)
