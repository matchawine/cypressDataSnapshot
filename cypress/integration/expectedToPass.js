import { dataSnapshotExpect } from "../../src/commands/dataSnapshotExpect"

describe("Data snapshot snapshot testing", function () {
  it("Test is passing", () => {
    cy.toMatchSnapshot({ test: true })
  })

  it("Test with property matcher", () => {
    cy.toMatchSnapshot(
      {
        test: true,
        anyDate: new Date(),
        anyNumber: 3,
        anyString: new Date().toString(),
        anyBoolean: true,
        anyArray: [true],
        anyRegExp: /yo/,
        anyObject: { test: true },
        stringMatching: "yolo",
        stringNotMatching: "yilo",
      },
      {
        anyDate: dataSnapshotExpect("any", Date),
        anyNumber: dataSnapshotExpect("any", Number),
        anyString: dataSnapshotExpect("any", String),
        anyBoolean: dataSnapshotExpect("any", Boolean),
        anyArray: dataSnapshotExpect("any", Array),
        anyRegExp: dataSnapshotExpect("any", RegExp),
        anyObject: dataSnapshotExpect("any", Object),
        stringMatching: dataSnapshotExpect("stringMatching", /^yo/),
        stringNotMatching: dataSnapshotExpect("not.stringMatching", /^yo/),
      },
    )
  })
})
