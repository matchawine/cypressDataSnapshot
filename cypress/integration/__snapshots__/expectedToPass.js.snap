// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data snapshot snapshot testing Test is passing 1`] = `
Object {
  "test": true,
}
`;

exports[`Data snapshot snapshot testing Test with property matcher 1`] = `
Object {
  "anyArray": Any<Array>,
  "anyBoolean": Any<Boolean>,
  "anyDate": Any<Date>,
  "anyNumber": Any<Number>,
  "anyObject": Any<Object>,
  "anyRegExp": Any<RegExp>,
  "anyString": Any<String>,
  "stringMatching": StringMatching /\\^yo/,
  "stringNotMatching": StringNotMatching /\\^yo/,
  "test": true,
}
`;
